Added MTB_transmission tutorial (#3211)
galaxyproject/training-material: topics/variant-analysis/tutorials/MTB_transmission/tutorial.md

> > is composed by two samples linked by recent transmission and transmission cluster 12 by three samples
> > linked by recent transmission. For example samples ERR6362484 and ERR5987352 are linked by
> > recent transmission.
> {: .solution}
{: .question}

> ### {% icon question %} Question
>
> 1. Let's assume that we have the isolation dates of samples ERR6362484 and ERR5987352, which
> belong to the same transmission cluster. Sample ERR6362484 was isolated on January 2021, while sample
>  ERR5987352 was isolated on September 2021. Would you be able to determine who was the infector and who the infectee?
>
> > ### {% icon solution %} Solution
> > 1. **NO**
> >
> >  Isolation dates have been used traditionally to define **index cases** within transmission clusters
> > under the assumption that the most likely scenario is the first isolated sample to be the
> > source of transmission. Today we know that this assumption often leads
> > to misidentification of index cases. **Remember:** we cannot rule out the possibility
> > that patients within the cluster were infected by an index case that was not sampled.
> >

> > >  Read [Xu et al., 2019](https://doi.org/10.1371/journal.pmed.1002961) for more information on this topic.

Reviewer: This should be a citation

> > some of these patients were infected with an MDR strain somewhere else.
> >
> > When looking at clustered strains, distinguishing between transmitted and *de-novo* may be tricky. Note
> > that, for example, for the two RIF-monoresistant strains linked within the same transmission
> > cluster there are, at least, a couple of scenarios possibe: one in which a RIF-monoresistant strain
> > evolved in one patient and was transmitted to the other patient afterwards, and one in which both
> > were infected with the same RIF-monoresistant strain from a third patient that we have not sampled.
> > We need to note that, in the first scenario, drug resistance evolved *de novo* in one patient,
> > and was *transmitted* to the other patient, whereas in the second scenario drug resistance was
> > *transmitted* in both cases.
> {: .solution}
{: .question}

> ### {% icon question %} Question
>
> 1. The same principles than those explained above apply to the three MDR strains that are
> within the same transmission cluster. However in this case there is one strain that shows clear
> evidence of *de-novo* evolution of DR. Do you know which strain and why?

> > 2. Are there possible scenarios other than *de-novo* evolution of DR for this strain?

Reviewer: is this question unanswered? I only see a 1 below

Author: Yes you are right. I have removed this question as it did not add that much.

- Calculate pairwise SNP distances between MTB samples
- Identify transmission clusters based on SNP distances
- Study the emergence and spread of drug resistance based on transmission analysis.
time_estimation: 2H
enable: true
level : Intermediate
key_points:
- Clustering is a useful tool to detect transmission links between patients and oubreak investigation.
- Clustering can be used to investigate the transmission of certain traits, like drug resistance.
- Clustering does not provide information about particular transmission events nor their directionality (who infected whom).
- Clustering is very much influenced by sampling. Lower sampling proportions and shorter sampling timeframes lead to lower clustering rates that shoud not be confounded with lack of transmission.
contributors:
- GaloGS
- dbrites
- cstritt
---

# Pre-requisites
This tutorial belongs to *Day 3: Evolutionary epidemiology: using phylogenetics to understand DR emergence and Mtb transmission* of the workshop [Mycobacterium tuberculosis NGS made easy: data analysis step-by-step](https://gallantries.github.io/video-library/events/mtb-ngs/program.html). It assumes you went through the materials

> of *Day 2*, and you have watched the respective webinars and completed the tutorial on [MTB variant analysis](https://training.galaxyproject.org/training-material/topics/variant-analysis/tutorials/tb-variant-analysis/tutorial.html).

Reviewer: We generally avoid mentioning events things are part of, these could be better expressed through pre-requisites (see other tutorials)

The way you can perform the same analysis tasks on multiple samples at a a time in Galaxy (being it
a simple step like running Trimmomatic, or a complete analysis pipeline composed of many different steps)
is by using `Dataset collections`. To know more about dataset collections,
you can [have a look at the following material](https://galaxyproject.org/tutorials/collections/), however the idea is very simple:
all samples to be analyzed are merged into a "dataset collection", and then you analyze that "collection"
as you would analyze a single sample, with the difference that all steps will be performed for each
sample individually. Do not worry if you do not fully understand this right now, we will be creating
a dataset collection at the beginning of this tutorial and we will analyze it.

To save you some time, (and server load) we have run the pipeline that you used in the previous tutorial for
the 20 samples that we want to analyze. Thus, we now have 20 VCF files that describe the
mutations found for each of the samples. **These 20 VCFs files will be the starting point of this tutorial.**
 If you want to perform the mapping and variant calling for

> all of the samples, feel free to do it. You can find the respective [FASTQ files here](https://zenodo.org/record/5911437), and

Reviewer: This should be in the zenodo_link at top I think

Author: This is the only thing I am not sure about. For the workshop/tutorial we have created two repositories. One only has the fastq files (because this repository will be also shared and used for other purposes). The other one has the analysis "results", so only the VCFs and the output of TB-profiler. The zenodo link at top points to the second one, while in this paragraph we point to a second, optional, Zeonodo repository.

> the [Galaxy workflow that was used to analyze the samples here](https://usegalaxy.eu/u/galo_a_goig/w/from-fastqs-to-vcfs-and-bams).

Reviewer: if you have workflows, it's best to just include them in the repo, and point to those so people can run them anywhere, rather than pointing a worklfow on a specific server.

> - You can also [watch this one minute video](https://www.youtube.com/watch?v=F3qNs1_675g) about how to create collections
{: .tip}

# Generate a SNP alignment

In this tutorial we aim to calculate the genetic differences (in SNPs) between pairs of MTB genomes.
To do so, we need to compare, between each pair of genomes (thus pairwise), the nucleotides that are
observed at each position. Each time we find a different nucleotide at a given position, we will sum
1 SNP of genetic distance. For example, if two strains have a genetic distance of 5 SNPs, that would
mean that their respective genomes are almost identical, except for 5 positions along all the genome
in which they have different nucleotides.

To do such calculation we need to first build an alignment of all the genomes (multiple-sequence
  alignment, or MSA). Afterwards, we will use specific software to analyze this MSA, count SNPs,
  and thus calculate the genetic distance between each pair of samples.

![MSA of 5 MTB genomes](./images/MSA.png "A SNP is highlighted in a MSA of five MTB genomes")

## Generate complete genomes
The first step to generate the genomes MSA will be... to get the complete genomes of our samples!

> In the [MTB Variant Analysis tutorial](https://training.galaxyproject.org/training-material/topics/variant-analysis/tutorials/tb-variant-analysis/tutorial.html) we have analyzed short-read high-throughput sequencing data (Illumina) to

Reviewer: please use the link command everywhere

```suggestion
In the [MTB Variant Analysis tutorial]({% link topics/variant-analysis/tutorials/tb-variant-analysis/tutorial.md %}) we have analyzed short-read high-throughput sequencing data (Illumina) to
```

| Sample C | MDR        |
| Sample Z | XDR        |

If we have a look at a TB profiler report we can see that there is one line describing the genotypic drug susceptibility.
Let's have a look at the first part of the TB-profiler report for sample ERR6362653:

```

TBProfiler report
=================

The following report has been generated by TBProfiler.

Summary
-------
ID: tbprofiler
Date: Fri Jan 28 13:14:47 2022
Strain: lineage2.2.1
Drug-resistance: MDR

Lineage report
--------------
Lineage	Estimated Fraction	Family	Spoligotype	Rd
lineage2	1.000	East-Asian	Beijing	RD105
lineage2.2	0.996	East-Asian (Beijing)	Beijing-RD207	RD105;RD207
lineage2.2.1	0.999	East-Asian (Beijing)	Beijing-RD181	RD105;RD207;RD181
```

As you can see, this strain is multi-drug resistant as inficated by (`Drug-resistance: MDR`)
We could then look for this information in each TB-profiler report and generate this table manually,
for example in a spreadsheet. However this is not feasible when analyzing hundreds or thousands of samples
(and very **error-prone!**).

We are here to learn bioinformatics, so let's generate this table using Linux commands.

The process will consist on three steps:

> * 1) Select the line containing the drug resistance profile with **grep**:

Reviewer: this will not be a proper numbered list, please use 
```suggestion
1. Select the line containing the drug resistance profile with **grep**:
```
and then check the indentation of the child elements below (generally three spaces to match the text 'Select')

```
Drug-resistance: MDR
```

* 2) Prepend the name of the sample with **Add input name as column**:

```
ERR6362653.txt Drug-resistance: MDR
```

* 3) Concatenate results from all samples in a single file with **Concatenate datasets**:

```
ERR6362653.txt Drug-resistance: MDR
ERR313115.txt Drug-resistance: Sensitive
ERR5987300.txt Drug-resistance: Pre-XDR
.... etc
```

* 4) As an optional step, we can reformat the table with **sed** to get rid of the `.txt` and `Drug-resistance:`
so the table looks like:

```
ERR6362653 MDR
ERR313115 Sensitive
ERR5987300 Pre-XDR
```

> #### 1. Select the line containing the drug resistance profile with `grep`

Reviewer: ```suggestion
#### Select the line containing the drug resistance profile with `grep`
```
Please avoid numbers in section headings

Author: Hi, thanks a lot for the review! I have incorporated your suggestions, with few exceptions as commented above :)

---
layout: tutorial_hands_on

title: 'Identifying tuberculosis transmission links: from SNPs to transmission clusters'
zenodo_link: ''
objectives:
- Create a SNP alignment
- Calculate pairwise SNP distances between MTB samples
- Identify transmission clusters based on SNP distances
- Study the emergence and spread of drug resistance based on transmission analysis.
time_estimation: 2H

> enable: true

Reviewer: ```suggestion
```

level : Intermediate
key_points:
- Clustering is a useful tool to detect transmission links between patients and oubreak investigation.
- Clustering can be used to investigate the transmission of certain traits, like drug resistance.
- Clustering does not provide information about particular transmission events nor their directionality (who infected whom).
- Clustering is very much influenced by sampling. Lower sampling proportions and shorter sampling timeframes lead to lower clustering rates that shoud not be confounded with lack of transmission.
contributors:
- Galo A. Goig
- Daniela Brites

> - Christoph Stritt

Reviewer: These three people need to be added to the `CONTRIBUTORS.yaml` file, like you did in https://github.com/galaxyproject/training-material/pull/3210, and then you should use their 'username' rather than their human name, to reference the entry in the CONTRIBUTORS.yaml file.

Author: Thank you, I will do it now

Author: Just one thing that is not clear to me: I added the contributors and commited the change to the main branch in my forked repository. Is this automatically picked up by this pull request or do I need to create a new one?

Reviewer 2: Yes, it is. This looks good. The only thing that you can improve next time is, to not commit anything to your main branch. Always use a new branch :)

---

# Introduction
{:.no_toc}

> Now you are familiar with the process of genome sequencing, quality control of sequencing data, mapping and

Author: Yes exactly. So the idea is definitely to have the link, but we are waiting to put everything together so we have also the final links and we can polish these kind of details! :)

> Now you are familiar with the process of genome sequencing, quality control of sequencing data, mapping and

Reviewer 3: You can always point to existing GTN tutorials on those topics here, for anybody finding this tutorial not in the context of your training course next week. :+1:

relationships, and for these reasons we build MSA for sequences for which we already assume some sort
of evolutionary relationship. More on MSAs here (link here to a webpage or Christoph's tutorial.).
**Galo: maybe this is too much, or too complex explanation or not necessary here**

We can build MSAs of a particular gene, a particular sequence or the complete genome. Building MSAs
of several complete genomes can be a complicated process and computationally demanding. To perform
such task there are many software packages available like `Muscle`, `MAFFT` or `Clustal` just to mention some.
Which one are we going to use in this tutorial? Well, we are going to use a trick. We are going to
just stack one genome on top of each other within a text file. (More on why we can do this below).

Our aim is to generate a **multifasta** file in which the genomes of our samples are aligned.
Something that looks like this:

```
>Sample 1
TTGACCGATGACCCCGGTTCAGGCTTCACCACAGTGTGGAACGC...
>Sample 2
ATGACCGATGACCCCGGTTCAGGCTTCACCACAGTGTGGAACGC...
>Sample 3
TTCGTTCGATGACCCCGGTTCAGGCTTCACCACAGTGTGGAACG...
```

> But given the output of **bcftools consensus**, what we have **for each** sample looks like this:

Author: Oh I missed that, that would indeed make things waay less complicated. Thanks a lot!

patients are within the same transmission cluster or not. Having a SNP alignment, this is fairly
easy. We will use **SNP distance matrix**, that will generate a matrix with pairwise SNP distances.

> ### {% icon hands_on %} Hands-on: Distance matrix from SNP alignment.
>
> 1. {% tool [SNP distance matrix](toolshed.g2.bx.psu.edu/repos/iuc/snp_dists/snp_dists/0.6.3+galaxy0) %} with the following parameters:
>    - {% icon param-file %} *"FASTA multiple sequence alignment"*: `Single dataset`
> (output of **Finds SNP sites** {% icon tool %})
>
>    > ### {% icon comment %} Comment
>    >
>    > Have a look at the distance matrix to make sure you understand the whole process. Given that
>    > we only have 20 samples, you could already spot some samples that are involved in the same
>    > transmission chain (samples with a small number of SNPs between them as explained below).
>    {: .comment}
>
{: .hands_on}

## Determine transmission clusters based on a SNP threshold
Now that we have a distance matrix that describes the SNP distance between each pair of samples, we
could already describe the transmission clusters based on a **SNP threshold**, as explained in the

> respective webinar(link). If two samples are at a distance below that threshold, we will say that they

Author: Of course I do! ;)  The webinars we will be uploading them to YouTube this week presumably. I will link them as soon as we have them!

> >
> > 1. The figure used above as an example is a **flagrant oversimplification**. In the figure not all pairwise
> > distances are represented (for example between sample A and C).
> > Most importantly you have to remember that **transmission clusters do not reflect transmission
> > events**.
> > In fact it may happen that transmission does not happen within the cluster we are analyzing!
> > For example, when a patient that is not sampled is the source of infection of all the cases in
> > the cluster (may be a superspreading event). You have to consider that, taking into account the
> > same SNP distances, another possible (yet still oversimplified) scenario could be...
> > ![MSA of 5 MTB genomes](./images/Cluster2.png "Example of transmission cluster using 10 SNPs threshold")
> {: .solution}
{: .question}

## Determine transmission clusters using Rscript
Currently there is not tool in Galaxy to perform the exact task that we need, but we can use the
statistics package `R` and the library `cluster` to perform such task. Again, don't worry about this,
programming in `R` is beyong the scope of this workshop. We have prepared a simple `R` script that

> identifies transmission clusters based on the matrix of SNP distances. This script is available

Reviewer 4: Ok... it might make sense to make a little tool? If you need help with the Galaxy tool side of things just shout.

> > a look at the TB profiler report, we can see that this strain carries an additional mutation in
> > *gyrA* that confers resistance to fluorioquinolones. This is compatible with an scenario in which
> > fluoroquinolone resistance evolved independently within this patient after being infected with
> > the MDR strain.
> > 2. Remember, although it is reasonable to think that fluorioquinolone resistance evolved in this
> > patient, we cannot rule out that actually it evolved in another patient who we did not sample and
> > was the transmitter of the fluoroquinolone-resistant strain.
> {: .solution}
{: .question}

> ### {% icon question %} Question
>
> 1.  There is one strain with a DR profile "other", because it is only resistant to pyrazinamide. This
> strain is not within a transmission cluster. Therefore, we conclude that pyrazinamide resistance
> most likely evolved *de-novo* in this patient. But we are wrong. Do you know why?  
>
> > ### {% icon solution %} Solution
> > 1. The strain is indeed PZA-resistant. And indeed this is strain is NOT linked to transmission
> > within our population. However, if we have a look at the TB-profiler report, we observe that this

> > > is a *M. bovis* strain, wich are known to be intrinsically resistant to PZA (all carry the same

Reviewer 4: typo: which
